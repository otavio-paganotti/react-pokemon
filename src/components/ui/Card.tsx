import { PokemonItem } from "@/types";
import clsx from "clsx";
import { ReactNode } from "react";
import { Link } from 'react-router-dom';

interface IProps<T> {
  children?: ReactNode;
  item: T;
  noLink?: boolean;
}

const Card = <T extends PokemonItem>({ item, children, noLink }: IProps<T>) => {
  const getHP = () => item.content.stats?.find(stat => stat.stat.name === 'hp')?.base_stat;
  const getAttack = () => item.content.stats?.find(stat => stat.stat.name === 'attack')?.base_stat;
  const getDefense = () => item.content.stats?.find(stat => stat.stat.name === 'defense')?.base_stat;

  return (
    <div className="bg-white">
      <div className="flex justify-center">
        <img src={`https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/${item.id}.png`} />
      </div>

      <div className="p-4 flex flex-col gap-4">
        <div className="bg-primary">
          <div className="p-2 flex justify-between">
            <h2 className="font-bold capitalize">{item.name}</h2>
            <span>HP: {getHP()}</span>
          </div>
          <div className={clsx("px-2 pt-2", !noLink ? "pb-2" : "")}>
            <p>Ataque: {getAttack()}</p>
            <p>Defesa: {getDefense()}</p>
          </div>
          {children}
        </div>
        {!noLink && (<div className="flex justify-center">
          <Link to={`/${item.id}`} className="bg-primary text-white px-4">Access Pokemon</Link>
        </div>)}
      </div>
    </div>
  )
}

export default Card
