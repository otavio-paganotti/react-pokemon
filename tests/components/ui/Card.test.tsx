import React from 'react';
import renderer from 'react-test-renderer';
import Card from '@/components/ui/Card';

const toJson = (component: renderer.ReactTestRenderer) => {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as renderer.ReactTestRendererJSON
};

test('should render correctly the card component', () => {
  const component = renderer.create(
    <Card>
      This is a content
    </Card>
  );

  let tree = toJson(component);
  expect(tree).toMatchSnapshot();

  expect(tree.children).toHaveLength(2);
  expect(tree.children[1].children[0]).toBe('This is a content');
});
