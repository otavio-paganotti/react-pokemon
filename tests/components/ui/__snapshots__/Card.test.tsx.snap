// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render correctly the card component 1`] = `
<div>
  this is a card
  <div>
    This is a content
  </div>
</div>
`;
